libs/cardano-ledger-core/src/Cardano/Ledger/DRepDistr.hs

```diff
@@ -0,0 +1,222 @@
+import Control.Monad.Identity (Identity (..))
+import Data.Kind
+import Data.Pulse (Pulsable (..), complete, pulse)
+import Data.Typeable
+import Data.VMap (VB, VMap (..), VP, foldlWithKey)
+import qualified Data.VMap as VMap
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks (..), allNoThunks)
+
+-- import Cardano.Ledger.EpochBoundary (SnapShot (..),  Stake (..))
+
+-- =================================================================
+-- Algorithm for computing the DRep stake distrubution
+
+-- | Given three inputs
+--   1) Map (Credential 'Staking c) (DRep c).   The delegation map. Inside the DRepUView of the UMap 'um' from the DState.
+--   2) regDreps :: Set (Credential 'Voting c). The first part of the VState.
+--   3) stakeDistr :: VMap VB VP (Credential 'Staking c) (CompactForm Coin). The first part of the Mark SnapShot.
+--   Compute the Drep distribution of stake(Coin)
+--  cost is expected to be O(size of 'stakeDistr' * log (size of 'um') * log (size of 'regDreps'))
+--  This is going to be expensive, so we will want to pulse it. Without pulsing, we estimate 3-5 seconds
+drepDistr ::
+  UMap c ->
+  Set (Credential 'Voting c) ->
+  VMap VB VP (Credential 'Staking c) (CompactForm Coin) ->
+  Map (DRep c) (CompactForm Coin)
+drepDistr um regDreps stakeDistr = foldlWithKey (accumDRepDistr um regDreps) Map.empty stakeDistr
+
+accumDRepDistr ::
+  UMap c ->
+  Set (Credential 'Voting c) ->
+  Map (DRep c) (CompactForm Coin) ->
+  Credential 'Staking c ->
+  (CompactForm Coin) ->
+  Map (DRep c) (CompactForm Coin)
+accumDRepDistr um regDreps ans stakecred c =
+  case UMap.lookup stakecred (DRepUView um) of
+    Nothing -> ans
+    Just drep@DRepAlwaysAbstain -> Map.insertWith UMap.addCompact drep c ans
+    Just drep@DRepAlwaysNoConfidence -> Map.insertWith UMap.addCompact drep c ans
+    Just drep@(DRepCredential cred2) ->
+      if Set.member cred2 regDreps
+        then Map.insertWith UMap.addCompact drep c ans
+        else ans
+
+-- =================================================================
+-- Making the algorithm Pulseable
+
+-- | The type of a Pulser which uses 'accumDRepDistr' as its underlying function.
+--   'accumDRepDistr' will be partially applied to the components of type (UMap c)
+--   and (Set (Credential 'Voting c)) when pulsing. Note that we use two type
+--   equality (~) constraints to fix both the monad 'm' and the 'ans' type, to
+--   the context where we will use the type as a Pulser. The type DRepPulser must
+--   have 'm' and 'ans' as its last two parameters so we can make a Pulsable instance.
+--   We will always use this instantiation (DRepPulser c Identity (Map (DRep c) (CompactForm Coin)))
+data DRepPulser c (m :: Type -> Type) ans where
+  DRepPulser ::
+    (ans ~ Map (DRep c) (CompactForm Coin), m ~ Identity) =>
+    !Int ->
+    !(UMap c) ->
+    !(Set (Credential 'Voting c)) ->
+    -- purposely not strict
+    (VMap VB VP (Credential 'Staking c) (CompactForm Coin)) ->
```

Comment:
Why is it lazy? We force this field at every pulse, so it seems superfluous to specifically keep it lazy.

Reply:
If it was strict, it would be forced when we constructed the DRepDistr (at the epoch boundary) where we definitely don't want to incur this cost.  In fact this is the same thunk that the TICKF (forecast) rule forces.  It will be force by the first pulse, or the first use of TICKF, and then never again.